projects/bootloader/test/test_dispatcher.c

```diff
@@ -0,0 +1,263 @@
+    .type = is_start_message ? CAN_MSG_TYPE_ACK : CAN_MSG_TYPE_DATA,
+    .dlc = len,
+    .data = 0,
+  };
+  // copy the message data over to the message to be transmitted
+  memcpy(message.data_u8, data, len);
+  fifo_push(&s_tx_fifo, &message);
+  return STATUS_CODE_OK;
+}
+
+static void prv_tx_from_fifo(void) {
+  CanMessage message;
+  fifo_pop(&s_tx_fifo, &message);
+  can_transmit(&message, NULL);
+}
+
+static void prv_mock_tx_compl(void) {
+  s_tx_cmpl = true;
+}
+
+static void prv_send_tx_to_fifo(CanDatagramTxConfig tx_config) {
+  fifo_init(&s_tx_fifo, s_tx_buffer);
+  s_tx_cmpl = false;
+
+  s_tx_cmpl_cb = tx_config.tx_cmpl_cb;
+  tx_config.tx_cmpl_cb = prv_mock_tx_compl;
+  tx_config.tx_cb = prv_tx_to_fifo;
+  can_datagram_start_tx(&tx_config);
+
+  Event e = { 0 };
+  while (!s_tx_cmpl) {
+    MS_TEST_HELPER_AWAIT_EVENT(e);
+    can_datagram_process_event(&e);
+    can_process_event(&e);
+  }  // puts all messages that are "sent" into the fifo
+  TEST_ASSERT_EQUAL(DATAGRAM_STATUS_TX_COMPLETE, can_datagram_get_status());
+  s_tx_cmpl_cb();
+}
+
+void setup_test(void) {
+  event_queue_init();
+  interrupt_init();
+  soft_timer_init();
+  crc32_init();
+
+  bootloader_can_init(&s_test_can_storage, &s_test_can_settings, s_board_id);
+  can_datagram_init(&s_test_datagram_settings);
+}
+
+void teardown_test(void) {}
+
+StatusCode prv_dispatch_check_cb(uint8_t *data, uint16_t len, void *context) {
+  LOG_DEBUG("Callback called!\n");
+  bool *cb_called = context;
+  *cb_called = true;
+
+  memcpy(s_rx_data, data, len);
+  s_rx_data_len = len;
+  return STATUS_CODE_OK;
+}
+
+void test_dispatch_calls_cb(void) {
+  // tests basic dispatch calls the callback when datagram is recieved
+  // and tests dispatch's tx_cmpl_cb resumes datagram rx correctly
+  TEST_ASSERT_OK(dispatcher_init(s_board_id));
+
+  CanDatagramTxConfig tx_config = {
+    .dgram_type = TEST_DATA_GRAM_ID,
+    .destination_nodes_len = 1,
+    .destination_nodes = &s_board_id,
+    .data_len = 0,
+    .data = NULL,
+    .tx_cmpl_cb = tx_cmpl_cb,
+  };
+  prv_send_tx_to_fifo(tx_config);
+  // datagram is now in s_tx_fifo
+
+  bool cb_called = false;
+  TEST_ASSERT_OK(
+      dispatcher_register_callback(TEST_DATA_GRAM_ID, prv_dispatch_check_cb, &cb_called));
+
+  Event e = { 0 };
+  prv_tx_from_fifo();  // start msg (first msg in fifo)
+  MS_TEST_HELPER_CAN_TX_RX(CAN_DATAGRAM_EVENT_TX, CAN_DATAGRAM_EVENT_RX);
+
+  // datagram will be in rx mode while we send the datagram with can from fifo.
+  size_t tx_left = fifo_size(&s_tx_fifo);
+  bool send_tx = true;
+  while (can_datagram_get_status() == DATAGRAM_STATUS_ACTIVE) {
+    // this loop will loop through both command rx and response tx
+    if (send_tx && tx_left > 0) {
+      prv_tx_from_fifo();
+      send_tx = false;
+      tx_left--;
+    }
+    MS_TEST_HELPER_AWAIT_EVENT(e);
+    if (e.id == CAN_DATAGRAM_EVENT_TX) {
+      send_tx = true;
+    }
+    can_process_event(&e);
+    can_datagram_process_event(&e);
+  }
```

Comment:
When you go to make the datagram testing library, you'll want to factor this out into its own function so we're not copy/pasting it in every test. But that's for the next ticket.

Reply:
yep, the read into fifo, send from fifo, rx into fifo, and rx from fifo will all be factored out a seperate functions.